### ISSUE TITLE
Type guard on conditional parameter type

### ISSUE THREAD

**Issue by jer-sen** [microsoft/TypeScript#38863]
**TypeScript Version:**  3.9.2

**Search Terms:**

**Expected behavior:** No error, Typescript should guess that a variable of type `T extends '1' ? '1' : never` will never be `false`, so that `NonFalse<TypeWithCondition<T>>` is exactly the same than `true | (T extends '1' ? '1' : never)`

**Actual behavior:** Error on `logNonFalse(a);`

**Related Issues:** none

**Code**
```ts
type TypeWithCondition<T> = boolean | (T extends '1' ? '1' : never);

type NonFalse<T> = Exclude<T, false>;

function logNonFalse<T>(b: NonFalse<TypeWithCondition<T>>) {
    console.log(b);
};

function test1<T>(a: TypeWithCondition<T>) {
    if (a === false) throw new Error("Can't be false");
    logNonFalse(a);
    // Error : Argument of type 'true | ([T] extends ["1"] ? "1" : never)' is not assignable to parameter of type 'true'.
    //  Type '[T] extends ["1"] ? "1" : never' is not assignable to type 'true'.
    //    Type '"1"' is not assignable to type 'true'.(2345)
};
```

<details><summary><b>Output</b></summary>

```ts
"use strict";
function logNonFalse(b) {
    console.log(b);
}
;
function test1(a) {
    if (a === false)
        throw new Error("Can't be false");
    logNonFalse(a);
    // Error : Argument of type 'true | ([T] extends ["1"] ? "1" : never)' is not assignable to parameter of type 'true'.
    //  Type '[T] extends ["1"] ? "1" : never' is not assignable to type 'true'.
    //    Type '"1"' is not assignable to type 'true'.(2345)
}
;

```

</details>

<details><summary><b>Compiler Options</b></summary>

```json
{
  "compilerOptions": {
    "noImplicitAny": true,
    "strictNullChecks": true,
    "strictFunctionTypes": true,
    "strictPropertyInitialization": true,
    "strictBindCallApply": true,
    "noImplicitThis": true,
    "noImplicitReturns": true,
    "alwaysStrict": true,
    "esModuleInterop": true,
    "declaration": true,
    "experimentalDecorators": true,
    "emitDecoratorMetadata": true,
    "moduleResolution": 2,
    "target": "ES2017",
    "jsx": "React",
    "module": "ESNext"
  }
}
```

</details>

**Playground Link:** [Provided](https://www.typescriptlang.org/play?jsx=0&ts=3.9.2#code/C4TwDgpgBAKuEHUCWwAWBhA9gOwCYqRwB4YA+KAXigCNNMAbCAQ2ygB8oAKGKCAD2AQ8AZygByAIxioAfnFSoALijYIANwgAnAJQBuAFD7QkKADkcAMSb1hEEuSoBRPgGN6AV1x2YAGigAza1tSA31-d2wXYEJWekwAc3NsKxtvUk5qZSSU2xJ4ZDQsPAJiMlJtKABvfShaqBccYQYIADo4+Iy9fQBfUPDI6JwoQWFgCXtOJmU4SAKMHHxB7HsK6rqoJH8uJkoKKkDUirRNTAB3FQhzx00TzU4AInQWMWAaaAPbe6719uygiEm3zqAHpgVBrrclFAAIKaeLuAC2QlemC2xmgL007mgHE4AG0YABdXgCIS4UR4+4Se7EuRU+5Q1QaHTSJCibCYV5MYTCJDxbBMaiMYaYKBgJiaJhIwSaKCo4bwcTALEQMQtGogsGwRViAnE-iCERQSnU2lQemM9RaVnszlQbm8-mC4XAUXopUqtUa2qg9YzDH0m0qO0OvkCoXQV0KkyY7FqzgAJgAzAAWACs2h6uiAA)

**Comment by RyanCavanaugh** [comment]
We can't analyze narrowing like this without negated types (#29317). Even then this might require higher order reasoning than that
